pkg/kv/kvserver/gc/gc_random_test.go

```diff
@@ -183,6 +192,273 @@ func BenchmarkRun(b *testing.B) {
 	}
 }
 
+func TestNewVsInvariants(t *testing.T) {
+	ctx := context.Background()
+	const N = 100000
+
+	someVersionsMidSizeRowsLotsOfIntents := someVersionsMidSizeRows
+	someVersionsMidSizeRowsLotsOfIntents.intentFrac = 1
+	for _, tc := range []randomRunGCTestSpec{
+		{
+			ds: someVersionsMidSizeRowsLotsOfIntents,
+			// Current time in the future enough for intents to get resolved
+			now: hlc.Timestamp{
+				WallTime: (intentAgeThreshold + 100*time.Second).Nanoseconds(),
+			},
+			// GC everything beyond intent resolution threshold
+			ttlSec: int32(intentAgeThreshold.Seconds()),
+		},
+		{
+			ds: someVersionsMidSizeRows,
+			now: hlc.Timestamp{
+				WallTime: 100 * time.Second.Nanoseconds(),
+			},
+			ttlSec: 1,
+		},
+	} {
+		t.Run(fmt.Sprintf("%v@%v,ttl=%vsec", tc.ds, tc.now, tc.ttlSec), func(t *testing.T) {
+			rng := rand.New(rand.NewSource(1))
+			eng := storage.NewDefaultInMemForTesting()
+			defer eng.Close()
+
+			tc.ds.dist(N, rng).setupTest(t, eng, *tc.ds.desc())
+			snap := eng.NewSnapshot()
+
+			// Run GCer over snapshot.
+			ttl := time.Duration(tc.ttlSec) * time.Second
+			gcThreshold := CalculateThreshold(tc.now, ttl)
+			intentThreshold := tc.now.Add(-intentAgeThreshold.Nanoseconds(), 0)
+
+			gcer := makeFakeGCer()
+			gcInfoNew, err := Run(ctx, tc.ds.desc(), snap, tc.now,
+				gcThreshold, RunOptions{IntentAgeThreshold: intentAgeThreshold}, ttl,
+				&gcer,
+				gcer.resolveIntents,
+				gcer.resolveIntentsAsync)
+			require.NoError(t, err)
+
+			// Handle GC + resolve intents.
+			var stats enginepb.MVCCStats
+			require.NoError(t,
+				storage.MVCCGarbageCollect(ctx, eng, &stats, gcer.requests(), gcThreshold))
+			for _, i := range gcer.intents {
+				l := roachpb.LockUpdate{
+					Span:   roachpb.Span{Key: i.Key},
+					Txn:    i.Txn,
+					Status: roachpb.ABORTED,
+				}
+				_, err := storage.MVCCResolveWriteIntent(ctx, eng, &stats, l)
+				require.NoError(t, err, "failed to resolve intent")
+			}
+
+			assertGCInvariants(t, eng, snap, *tc.ds.desc(), tc.now, gcThreshold, intentThreshold, ttl,
+				gcInfoNew)
+		})
+	}
+}
+
+// assertGCInvariants will create a stream of expected values based on full data
+// set contained in provided "before" reader and compare it with the "after"
+// reader that contains data after applying GC request.
+func assertGCInvariants(
```

Comment:
nit: this isn't really asserting invariants as much as it's simulating GC. Gets the job done in any case, but it would be nice to explicitly list the invariants being simulated/asserted.